File: py/day02.py
import os

# https://adventofcode.com/2022/day/2 - Day 2: Rock Paper Scissors
# Why does modulo stuff mess with my brain so much

def points_for(x):
    return 1 if (x == "A" or x == "X") else 2 if (x == "B" or x == "Y") else 3

def get_score(a, b):
    pt_a = points_for(a)
    pt_b = points_for(b)
    win = pt_b % 3 == (pt_a + 1) % 3
    return pt_b + (3 if pt_a == pt_b else 6 if win else 0)

with open(os.path.dirname(os.path.realpath(__file__)) + "/../input/day02-input", "r") as f:
    input = [line.strip() for line in f.readlines()]

part1 = 0 # sum(get_score(p1, p2) for (p1, p2) in [(play[0], play[1]) for play in [i.strip().split(' ') for i in input]])
part2 = 0

for (p1, p2) in [(play[0], play[1]) for play in [i.split(' ') for i in input]]:
    part1 += get_score(p1, p2)
    pp1 = points_for(p1)
    if p2 == "X": # Lose
        part2 += 3 if pp1 == 1 else (pp1 - 1) # or ((pp1 - 1 - 1 + 3) % 3) + 1 == ((pp1 + 1) % 3) + 1
    elif p2 == "Y": # Draw
        part2 += 3 + pp1
    else: # Win
        part2 += 6 + (1 if pp1 == 3 else (pp1 + 1)) # or ((pp1 - 1 + 1 + 3) % 3) + 1 == (pp1 % 3) + 1

print("Part 1:", part1)
print("Part 2:", part2)

File: rs/src/days/day02.rs
use crate::common::*;

fn get_points(a: &str) -> i32 {
    match a {
        "A" | "X" => 1,
        "B" | "Y" => 2,
        "C" | "Z" => 3,
        _ => panic!("WTF")
    }
}

fn get_score(a: &str, b: &str) -> i32 {
    let pa = get_points(a);
    let pb = get_points(b);
    let is_win = pb % 3 == (pa + 1) % 3;

    if pa == pb { 3 + pb } else if is_win { 6 + pb } else { pb }
}

pub fn run() {
    let day = 2;
    let input = get_input_strs(&day).unwrap();
    let mut part1 = 0;
    let mut part2 = 0;
    for i in input {
        let spl: Vec<&str> = i.split(' ').collect();
        part1 += get_score(spl[0], spl[1]);

        // Part 2
        let pp1 = get_points(spl[0]);
        match spl[1] {
            "X" => part2 += ((pp1 + 1) % 3) + 1,  // Lose
            "Y" => part2 += 3 + pp1,              // Draw
            "Z" => part2 += 6 + ((pp1 % 3) + 1),  // Win
            _ => panic!("WTF")
        }
    }
    print_day_results(&day, &part1, &part2);
}